## LLM PEFT BERT with Context Relavant Co



In [1]:
# Let's import the public training set and take a look
import pandas as pd

train_df = pd.read_csv('./data/train.csv')
print(train_df.shape)
train_df.head()

(200, 8)


id                                             prompt  \
0   0  Which of the following statements accurately d...   
1   1  Which of the following is an accurate definiti...   
2   2  Which of the following statements accurately d...   
3   3  What is the significance of regularization in ...   
4   4  Which of the following statements accurately d...   

                                                   A  \
0  MOND is a theory that reduces the observed mis...   
1  Dynamic scaling refers to the evolution of sel...   
2  The triskeles symbol was reconstructed as a fe...   
3  Regularizing the mass-energy of an electron wi...   
4  The angular spacing of features in the diffrac...   

                                                   B  \
0  MOND is a theory that increases the discrepanc...   
1  Dynamic scaling refers to the non-evolution of...   
2  The triskeles symbol is a representation of th...   
3  Regularizing the mass-energy of an electron wi...   
4  The angular spacing of features in the diffrac...   

                                                   C  \
0  MOND is a theory that explains the missing bar...   
1  Dynamic scaling refers to the evolution of sel...   
2  The triskeles symbol is a representation of a ...   
3  Regularizing the mass-energy of an electron wi...   
4  The angular spacing of features in the diffrac...   

                                                   D  \
0  MOND is a theory that reduces the discrepancy ...   
1  Dynamic scaling refers to the non-evolution of...   
2  The triskeles symbol represents three interloc...   
3  Regularizing the mass-energy of an electron wi...   
4  The angular spacing of features in the diffrac...   

                                                   E answer  
0  MOND is a theory that eliminates the observed ...      D  
1  Dynamic scaling refers to the evolution of sel...      A  
2  The triskeles symbol is a representation of th...      A  
3  Regularizing the mass-energy of an electron wi...      C  
4  The angular spacing of features in the diffrac...      D

In [2]:
train_pages_df = pd.read_csv('./data/train_pages_df.csv')
phy_pages_df = pd.read_csv('./data/physics_pages_df.csv')

In [3]:
train_pages_df=pd.concat([train_pages_df,phy_pages_df],axis=0)

In [4]:
train_pages_df

page                     section_title  \
0     Supersymmetric quantum mechanics                      Introduction   
1     Supersymmetric quantum mechanics  Example: the harmonic oscillator   
2     Supersymmetric quantum mechanics  Example: the harmonic oscillator   
3     Supersymmetric quantum mechanics  Example: the harmonic oscillator   
4     Supersymmetric quantum mechanics  Example: the harmonic oscillator   
...                                ...                               ...   
4503               Negative resistance                       Oscillators   
4504               Negative resistance                        Amplifiers   
4505               Negative resistance                Switching circuits   
4506               Negative resistance                           History   
4507               Negative resistance                           Summary   

                                                   text  \
0     understanding consequences supersymmetry susy ...   
1     schr dinger equation harmonic oscillator takes...   
2     schr dinger equation harmonic oscillator takes...   
3     special case shape invariance, discussed below...   
4     special case shape invariance, discussed below...   
...                                                 ...   
4503  Negative differential resistance devices are w...   
4504  Negative differential resistance devices such ...   
4505  Negative differential resistance devices are a...   
4506  Negative resistance was first recognized durin...   
4507  In electronics, negative resistance (NR) is a ...   

                                           cleaned_text  chunk_id  
0     understanding consequences supersymmetry susy ...         1  
1     schr dinger equation harmonic oscillator takes...         1  
2     special case shape invariance, discussed below...         1  
3     schr dinger equation harmonic oscillator takes...         1  
4     special case shape invariance, discussed below...         1  
...                                                 ...       ...  
4503  negative differential resistance devices widel...         1  
4504  negative differential resistance devices gunn ...         1  
4505  negative differential resistance devices also ...         1  
4506  negative resistance first recognized investiga...         1  
4507  electronics, negative resistance nr property e...         1  

[5536 rows x 5 columns]

In [5]:
# For convenience we'll turn our pandas Dataframe into a Dataset
from datasets import Dataset
train_ds = Dataset.from_pandas(train_df)

In [6]:
train_ds

Dataset({
    features: ['id', 'prompt', 'A', 'B', 'C', 'D', 'E', 'answer'],
    num_rows: 200
})

In [7]:
train_ds[0]

{'id': 0,
 'prompt': 'Which of the following statements accurately describes the impact of Modified Newtonian Dynamics (MOND) on the observed "missing baryonic mass" discrepancy in galaxy clusters?',
 'A': 'MOND is a theory that reduces the observed missing baryonic mass in galaxy clusters by postulating the existence of a new form of matter called "fuzzy dark matter."',
 'B': 'MOND is a theory that increases the discrepancy between the observed missing baryonic mass in galaxy clusters and the measured velocity dispersions from a factor of around 10 to a factor of about 20.',
 'C': 'MOND is a theory that explains the missing baryonic mass in galaxy clusters that was previously considered dark matter by demonstrating that the mass is in the form of neutrinos and axions.',
 'D': 'MOND is a theory that reduces the discrepancy between the observed missing baryonic mass in galaxy clusters and the measured velocity dispersions from a factor of around 10 to a factor of about 2.',
 'E': 'MOND 

In [8]:
from transformers import AutoTokenizer
import pprint
tokenizer = AutoTokenizer.from_pretrained('bert-base-uncased')

In [9]:
# Now we'll instatiate the model that we'll finetune on our public dataset, then use to
# make prediction on the private dataset.
from transformers import AutoModelForMultipleChoice, TrainingArguments, Trainer
import torch
model = AutoModelForMultipleChoice.from_pretrained('bert-base-uncased')

Some weights of BertForMultipleChoice were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


In [10]:
import random
import pandas as pd
from nltk.tokenize import sent_tokenize

def prepare_nsp_data(corpus, max_samples=None):
    # Step 1: Split the corpus into sentences
    sentences = sent_tokenize(corpus)
    
    # Step 2: Create positive pairs (consecutive sentences)
    positive_pairs = [(sentences[i], sentences[i+1], 1) for i in range(len(sentences) - 1)]
    
    # Step 3: Create negative pairs (randomly picked second sentence)
    negative_pairs = []
    for i in range(len(sentences) - 1):
        sentence_A = sentences[i]
        random_sentence_B = random.choice(sentences)
        
        # Ensure sentence_B is not the actual next sentence
        while random_sentence_B == sentences[i + 1]:
            random_sentence_B = random.choice(sentences)
            
        negative_pairs.append((sentence_A, random_sentence_B, 0))
    
    # Combine positive and negative pairs
    dataset = positive_pairs + negative_pairs
    
    # Step 4: Optionally, limit the number of samples
    if max_samples:
        dataset = random.sample(dataset, max_samples)
    
    # Convert to a DataFrame for easy saving or manipulation
    df = pd.DataFrame(dataset, columns=["sentence_A", "sentence_B", "label"])
    
    return df

dfs=[]
for i in range(train_pages_df.shape[0]):
    df=prepare_nsp_data(train_pages_df['cleaned_text'].iloc[i])
    dfs.append(df)
combined_df = pd.concat(dfs, ignore_index=True)    
# Saving the dataset to a .csv file for BERT NSP training
#df.to_csv("nsp_training_data.csv", sep="\t", index=False)
print(combined_df.shape)
combined_df.head()

(91972, 3)


sentence_A  \
0  understanding consequences supersymmetry susy ...   
1  make progress problems, physicists developed s...   
2  hoped studying susy's consequences simpler set...   
3  example, students typically taught "solve" hyd...   
4  considerable amount work using many differenti...   

                                          sentence_B label  
0  make progress problems, physicists developed s...     1  
1  hoped studying susy's consequences simpler set...     1  
2  example, students typically taught "solve" hyd...     1  
3  considerable amount work using many differenti...     1  
4  outcome spectrum hydrogen atom energy states l...     1

In [11]:
import torch
from transformers import BertTokenizer
from torch.utils.data import Dataset, DataLoader

# Initialize the BERT tokenizer
tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')

# Custom Dataset class for BERT NSP task
class BertNSPDataset(Dataset):
    def __init__(self, dataframe, tokenizer, max_len):
        self.data = dataframe
        self.tokenizer = tokenizer
        self.max_len = max_len

    def __len__(self):
        return len(self.data)

    def __getitem__(self, index):
        # Get sentence A, sentence B, and label
        sentence_A = self.data.iloc[index]['sentence_A']
        sentence_B = self.data.iloc[index]['sentence_B']
        label = self.data.iloc[index]['label']
        
        # Encode sentence pair using BERT tokenizer
        encoding = self.tokenizer.encode_plus(
            sentence_A,
            sentence_B,
            add_special_tokens=True,       # Add [CLS], [SEP]
            max_length=self.max_len,       # Maximum length of the input sequence
            return_token_type_ids=True,    # Necessary for sentence pairs
            padding='max_length',          # Pad up to the max length
            truncation=True,               # Truncate sentences to the max length
            return_attention_mask=True,    # Create attention mask
            return_tensors='pt'            # Return PyTorch tensors
        )

        return {
            'input_ids': encoding['input_ids'].flatten(),
            'attention_mask': encoding['attention_mask'].flatten(),
            'token_type_ids': encoding['token_type_ids'].flatten(),
            'label': torch.tensor(label, dtype=torch.long)  # Convert label to tensor
        }

# Function to prepare the dataset
def create_nsp_dataset(df, tokenizer, max_len=128):
    return BertNSPDataset(df, tokenizer, max_len)

# Example of using the dataset
max_len = 128  # Max length of the input sequence

# Create the dataset
nsp_dataset = create_nsp_dataset(combined_df, tokenizer, max_len)

nsp_dataset_mini = create_nsp_dataset(combined_df.iloc[:128], tokenizer, max_len)

print('len of dataset:',len(nsp_dataset))
# Optionally, use DataLoader for batching
dataloader = DataLoader(nsp_dataset, batch_size=32, shuffle=True)


len of dataset: 91972


In [12]:
from transformers import BertForNextSentencePrediction, BertTokenizer, Trainer, TrainingArguments
from peft import PeftConfig, PeftModel
from peft import get_peft_model, LoraConfig, TaskType
import torch

# Initialize BERT with Next Sentence Prediction head
model = BertForNextSentencePrediction.from_pretrained('bert-base-uncased')

# Initialize the tokenizer
tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')

# Prepare PEFT configuration (for LoRA in this case)
peft_config = LoraConfig(
    task_type=TaskType.SEQ_CLS,  # NSP task
    r=8,                                         # Low-rank factor
    lora_alpha=32,                               # Scaling factor
    lora_dropout=0.1,                            # Dropout for LoRA
    target_modules=["query", "value"],           # Target attention layers
)

# Apply LoRA to the model
peft_model = get_peft_model(model, peft_config)

# Verify the trainable parameters
peft_model.print_trainable_parameters()

# Training arguments
training_args = TrainingArguments(
    output_dir="./data/peft/",       # Output directory
    evaluation_strategy="epoch",  # Evaluate at the end of every epoch
    learning_rate=5e-5,           # Learning rate
    per_device_train_batch_size=16,   # Batch size for training
    per_device_eval_batch_size=16,    # Batch size for evaluation
    num_train_epochs=1,           # Number of epochs
    weight_decay=0.001,            # Weight decay
    logging_dir='./logs',         # Directory for logging
    logging_steps=100,
)

# DataLoader already prepared
# `nsp_dataset` should be prepared as discussed in the previous steps

# Initialize the Trainer
trainer = Trainer(
    model=peft_model,             # PEFT BERT model with LoRA
    args=training_args,           # Training arguments
    train_dataset=nsp_dataset_mini,    # Training dataset
    eval_dataset=nsp_dataset_mini,     # Evaluation dataset (you can split the data if needed)
    tokenizer=tokenizer           # Tokenizer
)

# Start training
trainer.train()


/home/vino/anaconda3/envs/pytorch23/lib/python3.11/site-packages/transformers/training_args.py:1545: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
Be aware, overflowing tokens are not returned for the setting you have chosen, i.e. sequence pairs with the 'longest_first' truncation strategy. So the returned list will always be empty even if some tokens have been removed.
Be aware, overflowing tokens are not returned for the setting you have chosen, i.e. sequence pairs with the 'longest_first' truncation strategy. So the returned list will always be empty even if some tokens have been removed.
Be aware, overflowing tokens are not returned for the setting you have chosen, i.e. sequence pairs with the 'longest_first' truncation strategy. So the returned list will always be empty even if some tokens have been removed.
Be aware, overflowing tokens are not returned for the setting you have

trainable params: 294,912 || all params: 109,778,690 || trainable%: 0.2686


Be aware, overflowing tokens are not returned for the setting you have chosen, i.e. sequence pairs with the 'longest_first' truncation strategy. So the returned list will always be empty even if some tokens have been removed.
Be aware, overflowing tokens are not returned for the setting you have chosen, i.e. sequence pairs with the 'longest_first' truncation strategy. So the returned list will always be empty even if some tokens have been removed.
Be aware, overflowing tokens are not returned for the setting you have chosen, i.e. sequence pairs with the 'longest_first' truncation strategy. So the returned list will always be empty even if some tokens have been removed.
Be aware, overflowing tokens are not returned for the setting you have chosen, i.e. sequence pairs with the 'longest_first' truncation strategy. So the returned list will always be empty even if some tokens have been removed.
Be aware, overflowing tokens are not returned for the setting you have chosen, i.e. sequence pai

TrainOutput(global_step=8, training_loss=6.721007347106934, metrics={'train_runtime': 25.3182, 'train_samples_per_second': 5.056, 'train_steps_per_second': 0.316, 'total_flos': 8448544800768.0, 'train_loss': 6.721007347106934, 'epoch': 1.0})

In [17]:
### model trained in GPU. Load the trained model
### only 2 training epochs are done 
### to show model improvement over plain-BERT
model = AutoModelForMultipleChoice.from_pretrained('./data/peft/checkpoint-11000/')
tokenizer = AutoTokenizer.from_pretrained('./data/peft/checkpoint-11000/')

Some weights of BertForMultipleChoice were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


In [18]:
# We'll create a dictionary to convert option names (A, B, C, D, E) into indices and back again
options = 'ABCDE'
indices = list(range(5))

option_to_index = {option: index for option, index in zip(options, indices)}
index_to_option = {index: option for option, index in zip(options, indices)}

def preprocess(example):
    # The AutoModelForMultipleChoice class expects a set of question/answer pairs
    # so we'll copy our question 5 times before tokenizing
    first_sentence = [example['prompt']] * 5
    
    second_sentence = []
    for option in options:
        second_sentence.append(example[option])
    
    # Our tokenizer will turn our text into token IDs BERT can understand
    tokenized_example = tokenizer(first_sentence, second_sentence, padding='max_length', truncation=True, max_length=128)
    tokenized_example['label'] = option_to_index[example['answer']]

    return tokenized_example

tokenized_train_ds = train_ds.map(preprocess, batched=False, remove_columns=['prompt', 'A', 'B', 'C', 'D', 'E', 'answer'])

Map:   0%|          | 0/200 [00:00<?, ? examples/s]

In [20]:
# Function to make predictions
def predict_mcq(example, model):
    predictions = []
    
    # Extract necessary inputs from dataset
    input_ids = torch.tensor(example['input_ids']).unsqueeze(0)  # Add batch dimension
    attention_mask = torch.tensor(example['attention_mask']).unsqueeze(0)
    token_type_ids = torch.tensor(example['token_type_ids']).unsqueeze(0)
        
    # Get predictions
    with torch.no_grad():
        outputs = model(
            input_ids=input_ids,
            attention_mask=attention_mask,
            token_type_ids=token_type_ids
        )
        
    logits = outputs.logits
    predicted_index = torch.argmax(logits, dim=1).item()
    
    return predicted_index

# Run prediction over the dataset
pred_idx=[]
for data in tokenized_train_ds:
     pred_idx.append(predict_mcq(data , model))

In [21]:
# Get predicted answer
option_mapping = {0: 'A', 1: 'B', 2: 'C', 3: 'D', 4: 'E'}
prediction = [option_mapping[i] for i in pred_idx]

In [22]:
train_df['predicted_answer']=prediction

In [23]:
count=0
for i in range(train_df.shape[0]):
    if (train_df['answer'].iloc[i]==train_df['predicted_answer'].iloc[i]):
        count=count+1
print(count/200)

0.28


### BERT with PEFT (context data) improves from 0.18 to 0.28
#### context relavant corpus data converted to NSP daatset. Model only trained for 2 epochs on NSP style daatset. Further training may improve the accuracy.